### PR TITLE
[EX2] Create a module in Nest.js to handle user authentication using JWT (JSON Web Tokens).

### PR DESCRIPTION
**Pull request by er-anhlnt**
### **Summary**

- Implement api /auth/signup
- Implement api /auth/login: Login with email and password
- Implement api /auth/logout
- Implement api /auth/refresh
- Implement JWT Authguard
- Add [exercise3.md](https://github.com/er-anhlnt/nestjs-exercises/blob/EX3_Authentication_with_JWT/docs/exercise3.md)

### **Changes**
- Set JWT expiration to 15 minutes, Refresh token expiration is 6 months.
- Implement refresh token follow the [Refresh token rotation](https://auth0.com/docs/secure/tokens/refresh-tokens/refresh-token-rotation). Every time user send refresh token to get a new access token, we will create a new access token and a new refresh token.

**Automatic reuse refresh token detection:** 
1. Legitimate Client has refresh token 1, and it is leaked to or stolen by Malicious Client.
2. Legitimate Client uses refresh token 1 to get a new refresh token/access token pair.
3. Server returns refresh token 2/access token 2.
4. Malicious Client then attempts to use refresh token 1 to get an access token.  recognizes that refresh token 1 is being reused, and immediately invalidates the refresh token family, including refresh token 2.
5. Server returns an access denied response to Malicious Client.
6. Access token 2 expires and Legitimate Client attempts to use refresh token 2 to request a new token pair. Server returns an access denied response to Legitimate Client.
7. Re-authentication is required.
reuse-detection1

### **Checked for ensure quality**

/auth/login
login

/auth/signup
signup

/auth/logout
logout

/auth/refresh
refresh

/users/secret for testing jwt guard
secret

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,6 +1,10 @@
 NODE_ENV=development
+
+JWT_SECRET=eyijoijfij091238j18djas8djwq
```
Comment (thosanmaytinh) [inline .env:3]: e ko nên commit cả file .env lên như thế nào
e tạo 1 file .env.example chứa ví dụ thôi

File: src/auth/guards/auth.guard.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        throw new UnauthorizedException();
+      }
+
+      request.user = await this.jwtService.verifyAsync(token);
+
+      return true;
+    } catch (error) {
+      if (error instanceof TokenExpiredError) {
+        throw new UnauthorizedException('Token expired');
+      } else if (error instanceof JsonWebTokenError) {
```
Comment (thosanmaytinh) [inline src/auth/guards/auth.guard.ts:29]: cái này e không cần dùng tới else đâu, cứ 

if ...

if ...

if ... 

là nó throw rồi ko run tiếp step

File: db/migrations/1704249121152-create_user_and_refreshtoken_table.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class CreateUserAndRefreshtokenTable1704249121152
+  implements MigrationInterface
+{
+  name = 'CreateUserAndRefreshtokenTable1704249121152';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TABLE \`refresh_token\` (\`id\` int NOT NULL AUTO_INCREMENT, \`isActive\` tinyint NOT NULL DEFAULT 0, \`token\` varchar(255) NOT NULL, \`expiredAt\` datetime NOT NULL, \`createdAt\` datetime(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6), \`parent\` varchar(255) NULL, \`userId\` int NULL, UNIQUE INDEX \`IDX_c31d0a2f38e6e99110df62ab0a\` (\`token\`), PRIMARY KEY (\`id\`)) ENGINE=InnoDB`,
```
Comment (thosanmaytinh) [inline db/migrations/1704249121152-create_user_and_refreshtoken_table.ts:10]: tên bảng mình dùng số nhiều trong tiếng anh

refresh_tokens
users

File: .env (hunk 1 above)
Comment (er-anhlnt) [reply .env:3]: Dạ e hiểu r. E đã tạo PR mới để fix cái này ạ [Pull request URL ](https://github.com/er-anhlnt/nestjs-exercises/pull/8)

File: src/auth/guards/auth.guard.ts (hunk 2 above)
Comment (er-anhlnt) [reply src/auth/guards/auth.guard.ts:29]: Em sửa lại rồi a [Pull request URL](https://github.com/er-anhlnt/nestjs-exercises/pull/8)

File: db/migrations/1704249121152-create_user_and_refreshtoken_table.ts (hunk 3 above)
Comment (er-anhlnt) [reply db/migrations/1704249121152-create_user_and_refreshtoken_table.ts:10]: Em sửa lại rồi a [Pull request URL](https://github.com/er-anhlnt/nestjs-exercises/pull/8)

**Pull request merged**